File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        int row = matrix.length-1;
        int col = matrix[0].length-1;
        int sr = 0;
        int sc = 0;
        int c = 0, end = matrix.length*matrix[0].length;
        while(c<end)
        {
            for(int i=sc; i<=col && c<end; i++)
            {
                c++;
                ans.add(matrix[sr][i]);
            }
            sr++;
            for(int i=sr; i<=row && c<end; i++)
            {
                c++;
                ans.add(matrix[i][col]);
            }
            col--;
            for(int i=col; i>=sc && c<end; i--)
            {
                c++;
                ans.add(matrix[row][i]);
            }
            row--;
            for(int i=row; i>=sr && c<end; i--)
            {
                c++;
                ans.add(matrix[i][sc]);
            }
            sc++;
        }
        return ans;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []
        row = len(matrix) - 1
        col = len(matrix[0]) - 1
        sr = 0
        sc = 0
        c = 0
        end = len(matrix) * len(matrix[0])
        
        while c < end:
            for i in range(sc, col + 1):
                if c < end:
                    c += 1
                    ans.append(matrix[sr][i])
            sr += 1
            for i in range(sr, row + 1):
                if c < end:
                    c += 1
                    ans.append(matrix[i][col])
            col -= 1
            for i in range(col, sc - 1, -1):
                if c < end:
                    c += 1
                    ans.append(matrix[row][i])
            row -= 1
            for i in range(row, sr - 1, -1):
                if c < end:
                    c += 1
                    ans.append(matrix[i][sc])
            sc += 1
        
        return ans
